Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -225,9 +230,43 @@ public void run(@NotNull ProgressIndicator indicator) {
     @Override
     public void beforeDocumentChange(DocumentEvent e) {
       myServerData.onDocumentChanged(e);
+      synchronized (myLock) {
+        final Document document = e.getDocument();
+        myUpdatedDocuments.add(document);
+        scheduleUpdateFilesContent();
+      }
     }
   };
 
+  private void scheduleUpdateFilesContent() {
```

Comment:
That's the question to updateFilesContent() that iterates all unsaved docs and decides which ones to send.
The problem we are solving now is not to break ranges in files that we do already highlight using DAS. So this looks like a long term solution: once document is updated and it is visible at the moment make sure that on next IDE highlighting daemon pass it will be sent to the server (in order to get ranges, even if file is saved).